crates/wasmtime/src/runtime/types.rs

```diff
@@ -1676,13 +1696,9 @@ impl StructType {
     pub fn matches(&self, other: &StructType) -> bool {
         assert!(self.comes_from_same_engine(other.engine()));
 
-        // Avoid matching on structure for subtyping checks when we have
-        // precisely the same type.
-        if self.type_index() == other.type_index() {
-            return true;
-        }
```

Comment:
This is still a valid optimization to do though, right? IIRC `signatures()` below grabs a lock so ideally we'd keep this.

Reply:
`signatures()` doesn't grab a lock, the methods on it do, and the `is_subtype` method already has this optimization as a fast path, so no need to duplicate it all over the place.